File: src/aoc/2022/day-6/main.py
from pathlib import Path

def find_marker_start_position(line: str):
  package: list[str] = []
  for index, char in enumerate(line):
    if char in package:
      repeated_index = package.index(char)
      package = package[repeated_index + 1:]
    if len(package) == 4:
      return index, package
    package.append(char)
  return -1, package

def get_file_lines(file_path: str):
  file = Path(__name__).with_name(file_path)
  return file.read_text().splitlines()

def print_results(lines: list[str]):
	for index, line in enumerate(lines):
		position, package = find_marker_start_position(line)
		if position == -1:
			print(f'Unable to find Start Position and Package from line {index}')
		else:
			print(f'Start Position: {position} | Package: {package}')

FILE_PATH_EXAMPLE = "./example.txt"
FILE_PATH_INPUT = "./input.txt"

if __name__ == "__main__":
  lines = get_file_lines(FILE_PATH_EXAMPLE)
  print_results(lines)
  lines = get_file_lines(FILE_PATH_INPUT)
  print_results(lines)

File: src/aoc/2022/day-6/main.rb
def find_marker_position line
  package = []
  line.chars.each.with_index do |letter, index|
    if package.include?(letter)
      repeated_index = package.find_index(letter)
      package = package[repeated_index + 1..]
    end
    if package.length == 4
      return index, package
    end
    package.append(letter)
  end
  return -1, package
end

def get_file_lines file_path
  IO.readlines(file_path, chomp: true)
end

def print_results lines
  lines.each.with_index do |line, index|
		position, package = find_marker_position(line)
		if position == -1
			puts "Unable to find Start Position and Package from line #{index}"
		else
			puts "Start Position: #{position} | Package: #{package}"
    end
  end
end

FILE_PATH_EXAMPLE = "./example.txt"
FILE_PATH_INPUT = "./input.txt"

lines = get_file_lines(FILE_PATH_EXAMPLE)
print_results(lines)
lines = get_file_lines(FILE_PATH_INPUT)
print_results(lines)
